socorro-cmd

```diff
@@ -0,0 +1,199 @@
+        }
+    )
+]
+
+
+def import_and_run(runner):
+    """Takes a runner, converts it to a callable, and runs it
+
+    A runner can either be a Python-dotted string which will get imported and
+    run, or a callable which will get run directly.
+
+    """
+    if isinstance(runner, basestring):
+        # Import the class from the module
+        module_path, cls_name = runner.rsplit('.', 1)
+        __import__(module_path)
+        module = sys.modules[module_path]
+        app = getattr(module, cls_name)
+
+    else:
+        app = runner
+
+    # If the app is a main function, we run it as is
+    if isinstance(app, types.FunctionType):
+        sys.exit(app(sys.argv[1:]))
+
+    # If the app is a Socorro App, we run it using the Socorro App main
+    if issubclass(app, App):
+        sys.exit(app.run())
+
+    print('ProgrammerError: Unknown runner type')
+    sys.exit(1)
+
+
+def build_epilog():
+    """Builds the epilog containing the groups and commands"""
+    output = []
+    output.append('Available commands:')
+    output.append('\n')
+    for group in COMMANDS:
+        output.append('%s:' % group.name)
+        for cmd, runner in group:
+            output.append('  %s' % cmd)
+        output.append('\n')
+
+    output.append('')
+    output.append('Type "socorro-cmd CMD --help" for help on any command.')
+    output.append('')
+    output.append(
+        wrapped(
+            'All commands need to be run in a Docker container. Service commands are run in '
+            'the related container. For example, the "processor" command should run in the '
+            '"processor" container. All other commands should be run in the processor '
+            'container.'
+        )
+    )
+    output.append('')
+    output.append(
+        wrapped(
+            'For more documentation, see <https://socorro.readthedocs.io/>.'
+        )
+    )
+    return '\n'.join(output)
+
+
+def get_runner(cmd):
+    """Given a cmd, returns the runner"""
+    for group in COMMANDS:
+        runner = group.lookup(cmd)
+        if runner is not None:
+            return runner
+
+
+def cmd_main():
+    # Build a basic parser so we can take advantage of --help
+    parser = argparse.ArgumentParser(
+        prog='socorro-cmd',
+        formatter_class=argparse.RawDescriptionHelpFormatter,
+        epilog=build_epilog()
+    )
+    parser.add_argument(
+        'cmd',
+        nargs='?',
+        help='command to run'
+    )
+    parser.add_argument(
+        'arg',
+        nargs='*',
+        help='command arguments'
+    )
+
+    if not sys.argv[1:] or sys.argv[1] in ('-h', '--help'):
+        parser.print_help()
+        return 0
+
+    cmd = sys.argv[1]
+    runner = get_runner(cmd)
+    if runner is None:
+        parser.error('"%s" is not a valid command; see --help for command list' % cmd)
+
+    # Rewrite sys.argv so anything that parses sys.argv has the right one
+    sys.argv = [sys.argv[0] + ' ' + cmd] + sys.argv[2:]
```

Comment:
Maybe instead of modifying `argv` we can just pass around a truncated copy as arguments to the runners? Then we could avoid modifying global state. I'm not sure if it matters all that much though.

Reply:
This does two things:

First, argparse (which everything we've got uses) treats the first argument of sys.argv as the script and that's what it uses for `%PROG`. So the help text shows the right thing regardless of how you executed the script.

Second, the way `SocorroApp` is constructed, it lets configman do its thing in regards to command line arguments. We'd need to make non-trivial changes to `SocorroApp` and how it gets created. We could rework things, but given the first item, I decided not to at this juncture.

Reply:
For the record, I think that behavior of argparse is weird. It's overridable, but it's not conveniently so.

Reply:
Also, I'm not super wild about the way all this works. I think we'll hone it over time.